sympy/stats/symbolic_probability.py

```diff
@@ -0,0 +1,180 @@
+                if a.has(RandomSymbol):
+                    rv.append(a)
+            variances = Add(*map(lambda xv: Variance(xv, condition), rv))
+            map_to_covar = lambda x: 2*Covariance(*x, condition=condition)
+            covariances = Add(*map(map_to_covar, itertools.combinations(rv, 2)))
+            return variances + covariances
+        elif isinstance(arg, Mul):
+            nonrv = []
+            rv = []
+            for a in arg.args:
+                if a.has(RandomSymbol):
+                    rv.append(a)
+                else:
+                    nonrv.append(a**2)
+            if len(rv) == 0:
+                return S.Zero
+            if condition is None:
+                obj = Expr.__new__(cls, Mul(*rv))
+            else:
+                obj = Expr.__new__(cls, Mul(*rv), condition)
+            obj._condition = condition
+            return Mul(*nonrv)*obj
+
+        raise NotImplementedError
+
+    def doit(self, **kwargs):
+        return variance(self.args[0], self._condition, **kwargs)
+
+
+class Covariance(Expr):
+    """
+    Symbolic expression for the covariance.
+    """
+    def __new__(cls, arg1, arg2, condition=None, **kwargs):
+        arg1 = _sympify(arg1)
+        arg2 = _sympify(arg2)
+
+        if not kwargs.pop('evaluate', global_evaluate[0]):
+            if condition is None:
+                obj = Expr.__new__(cls, arg1, arg2)
+            else:
+                condition = _sympify(condition)
+                obj = Expr.__new__(cls, arg1, arg2, condition)
+            obj._condition = condition
+            return obj
+
+        if condition is not None:
+            condition = _sympify(condition)
+
+        if arg1 == arg2:
+            return Variance(arg1, condition)
+
+        if not arg1.has(RandomSymbol):
+            return S.Zero
+        if not arg2.has(RandomSymbol):
+            return S.Zero
+
+        arg1, arg2 = sorted([arg1, arg2], key=default_sort_key)
+
+        if isinstance(arg1, RandomSymbol) and isinstance(arg2, RandomSymbol):
+            return Expr.__new__(cls, arg1, arg2)
+
+        coeff_rv_list1 = cls._expand_single_argument(arg1.expand())
+        coeff_rv_list2 = cls._expand_single_argument(arg2.expand())
+
+        addends = [a*b*Covariance(r1, r2) for (a, r1) in coeff_rv_list1 for (b, r2) in coeff_rv_list2]
+        return Add(*addends)
+
+    @classmethod
+    def _expand_single_argument(cls, expr):
+        # return (coefficient, random_symbol) pairs:
+        if isinstance(expr, RandomSymbol):
+            return [(S.One, expr)]
+        elif isinstance(expr, Add):
+            outval = []
+            for a in expr.args:
+                rv = []
+                nonrv = []
+                if isinstance(a, Mul):
+                    outval.append(cls._get_mul_nonrv_rv_tuple(a))
+                elif isinstance(a, RandomSymbol):
+                    outval.append((S.One, a))
+
+                outval.append((Mul(*nonrv), Mul(*rv)))
```

Comment:
Under which conditions would ``nonrv`` or ``rv`` be non-empty? Is something missing?